### ISSUE TITLE
Nvidia GeForce RTX 3060 Ti always produces invalid shares

### ISSUE THREAD

**Issue by ahnick** [no-fee-ethereum-mining/nsfminer#100]
**Describe the bug**
I have a rig that is primarily GeForce GTX 1070s, but I just replaced one of the the cards with a GeForce RTX 3060ti (Ventus 2x OC) and I can't get it to produce valid shares.  It looks like it is mining (as I can see the hash rate as it is working), but every time it finds a solution I receive the message "gave incorrect result".  This happens even when using stock settings.   My 1070s on the same rig though, continue to produce valid shares.

I'm using NVIDIA Driver Version: 460.32.03 and CUDA Version: 11.2. 

**To Reproduce**
1. Build latest master branch from source using GCC 9.3 / CUDA 11.2 / driver 460.32
2. Launch nsfminer
3. Observe Error message that indicates the 3060ti GPU found a solution, but "gave incorrect result"

**Expected behavior**
No error message should be received and the card should be producing valid shares

**Environment (please complete the following information):**
- Operating System: Ubuntu 20.04
- Hardware [e.g. GTX 1070]: RTX 3060 Ti
- Ethminer Version [e.g. 0.14]: 1.3.5 (latest master)
- Ethminer options used: nsfminer -U -P stratum1+ssl://0x34e287e0f6f14449288caf7d52bff046f379069a.gpurig3@us1.ethermine.org:5555

**Additional context**
I have also tried mining with ethminer and T-Rex miner.  I get the same result with ethminer using the latest version and on T-Rex miner it gives the following error message even though I am using stock settings:

"cuda exception in [crc32_parallel, 146] illegal instruction was encountered, try to reduce overclock to stabilize GPU state"

**Comment by lesjokolat** [comment]
How much system memory do you have?  how big is linux swap? Perhaps increase system swap

**Comment by ahnick** [comment]
The rig has 4GB of system memory.  htop currently shows the system is utilizing 1.38G / 3.73.G.  Swap utilization is 0 / 3.74G.  nvidia-smi reports that the graphics card memory is 4400MiB / 7982MiB.  

Based on this information I don't believe I'm running into memory issues either on the system or the graphics card, so I don't think increasing the swap size will do anything as I am not observing any swap utilization currently on the machine.

**Comment by lesjokolat** [comment]
Technically looks ok. but I personally for 2 or more GPUs liked minimum 8 gigs ram. Would be interesting if you added 4 gigs more and tried.. Also in case some space needed for dag as its built then 4 gig ram would be tight... If you have the space add the swap double it from what it is now try it out and see.

**Comment by jean-m-cyr** [comment]
I don't have a 30xx gen GPU. I'd like to know if this is a generic problem on all 3060s, or is it specific to this setup?

**Comment by Minerx117** [comment]
i have 3060ti running at 61mhs no worrys at all 461.40

**Comment by tuero** [comment]
I have a 3090 and I will occasionally get the above (infinite loop after incorrect result). Will post a log if once I see it again.

**Comment by ahnick** [comment]
Tried setting the swap to 8G, but still always getting invalid shares.

```
$ swapon --show
NAME      TYPE SIZE USED PRIO
/swap.img file 3.8G   0B   -2
$ sudo swapoff /swap.img
$ sudo dd if=/dev/zero of=/swap.img bs=1M count=8192
8192+0 records in
8192+0 records out
8589934592 bytes (8.6 GB, 8.0 GiB) copied, 16.8275 s, 510 MB/s
$ ls -al /swap.img
-rw------- 1 root root 8589934592 Feb 19 16:26 /swap.img
$ sudo mkswap /swap.img
Setting up swapspace version 1, size = 8 GiB (8589930496 bytes)
no label, UUID=8a84dac6-805b-4b55-a530-460d6ff0c3bf
$ sudo swapon /swap.img
$ swapon --show
NAME      TYPE SIZE USED PRIO
/swap.img file   8G   0B   -2
```

I do think that the issue lies in an improperly generated DAG, as both t-rex miner and phoenix miner bailout during DAG generation due to errors coming from the card.  Ethminer and Nsfminer seem to be more resilient and generate a DAG, but if they are working from an incorrect DAG then that might explain why I'm always getting invalid shares.  I'm thinking it's an issue maybe with CUDA or the driver, but no idea how to prove that out.  Also, everyone I have heard from where the 3060 ti is working is running on a Windows setup.  If someone has a working 3060 Ti on Linux I would love to know the following:

OS version, NVIDIA driver version, CUDA version, GCC version, Miner and version

**Comment by sonicrules1234** [comment]
I've got a MSI GeForce RTX 3060 Ti 8 GB VENTUS 2X OC and it isn't producing any invalid shares.  If you overclock it wrong it will produce invalid shares, but that's not specific to nsfminer.

**Comment by ahnick** [comment]
@sonicrules1234 Per my original issue, the card is always producing invalid shares even at stock settings.  Are you running on Linux or Windows?  If Linux, what version of the NVIDIA driver and CUDA are you using?

**Comment by sonicrules1234** [comment]
> @sonicrules1234 Per my original issue, the card is always producing invalid shares even at stock settings. Are you running on Linux or Windows? If Linux, what version of the NVIDIA driver and CUDA are you using?

I'm running Manjaro Linux, nvidia driver version 460 and cuda version 11.2.

**Comment by ahnick** [comment]
@sonicrules1234 Okay, this is helpful.  

When the card is on stock settings, it will consume nearly all of the 200W that is available and the temperature of the card will heat up dramatically (80C+) and then the fan percentage will show "ERR" under nvidia-smi.  Does yours do anything like that?

I have some additional questions as well:
What version of nsfminer are you on?
Did you compile it from source?  If so, what GCC version did you use?
What settings are you using on your card?  (core clock offset, memory offset, power level)

Thanks!

**Comment by ahnick** [comment]
> @sonicrules1234 Okay, this is helpful.
> 
> When the card is on stock settings, it will consume nearly all of the 200W that is available and the temperature of the card will heat up dramatically (80C+) and then the fan percentage will show "ERR" under nvidia-smi. Does yours do anything like that?
> 
> I have some additional questions as well:
> What version of nsfminer are you on?
> Did you compile it from source? If so, what GCC version did you use?
> What settings are you using on your card? (core clock offset, memory offset, power level)
> 
> Thanks!

**Comment by sonicrules1234** [comment]
@ahnick I'm not sure I've ever used stock settings before.  At first I was just using the provided binary, but more recently I compile it myself. I'm on gcc version 10.2.0. As for overclocking, I'm using -500 core clock offset, 2200 memory offset, and 130W power limit.  I set my fans to 70%.

**Comment by ahnick** [comment]
> I've heard the nvidia-settings gui has some problems and only effects the first gpu. Lemme see if I can find the reddit post with how to do it all from the terminal.

@sonicrules1234 Yeah, I believe that a performance level 4 was introduced for the RTX 30XX series.  The reddit post makes the adjustment for all performance levels, but you can just use nvidia-settings from the command line to adjust performance level 4 by itself. 

I plugged in your settings below to the command I am using to set them:
nvidia-settings -c :0 -a '[gpu:0]/GPUGraphicsClockOffset[4]=-500' -a '[gpu:0]/GPUMemoryTransferRateOffset[4]=2200' -a '[gpu:0]/GPUFanControlState=1' -a '[fan:0]/GPUTargetFanSpeed=70'

Here is a sample of the resulting log file from mining:
```
Feb 27 14:02:12 gpurig3 bash[12392]:  miner 13:24 A1742:F580 202.59 Mh - cu0 59.82, cu1 28.21, cu2 28.65, cu3 28.45, cu4 28.39, cu5 29.07
Feb 27 14:02:14 gpurig3 bash[12392]:  miner Job: a3ca7802 blk: 11941377 us1.ethermine.org:5555
Feb 27 14:02:17 gpurig3 bash[12392]:  miner 13:24 A1742:F580 202.76 Mh - cu0 59.87, cu1 28.23, cu2 28.68, cu3 28.47, cu4 28.41, cu5 29.10
Feb 27 14:02:18 gpurig3 bash[12392]:  miner Job: 99207b8c blk: 11941377 us1.ethermine.org:5555
Feb 27 14:02:22 gpurig3 bash[12392]:  miner Job: bbe9659a blk: 11941377 us1.ethermine.org:5555
Feb 27 14:02:22 gpurig3 bash[12392]:  miner 13:24 A1742:F580 202.77 Mh - cu0 59.86, cu1 28.25, cu2 28.68, cu3 28.47, cu4 28.41, cu5 29.10
Feb 27 14:02:23 gpurig3 bash[12392]:  cu-0  Job: bbe9659a Solution: 0xe190708345ff7b13
Feb 27 14:02:23 gpurig3 bash[12392]:  miner GPU 0 gave incorrect result. Lower overclocking values if it happens frequently.
Feb 27 14:02:25 gpurig3 bash[12392]:  cu-0  Job: bbe9659a Solution: 0xe19070834baa1402
Feb 27 14:02:25 gpurig3 bash[12392]:  miner GPU 0 gave incorrect result. Lower overclocking values if it happens frequently.
Feb 27 14:02:26 gpurig3 bash[12392]:  miner Job: a431f8c6 blk: 11941377 us1.ethermine.org:5555
Feb 27 14:02:27 gpurig3 bash[12392]:  miner 13:24 A1742:F582 202.77 Mh - cu0 59.88, cu1 28.25, cu2 28.68, cu3 28.47, cu4 28.41, cu5 29.09
Feb 27 14:02:30 gpurig3 bash[12392]:  miner Job: 58185004 blk: 11941377 us1.ethermine.org:5555
Feb 27 14:02:30 gpurig3 bash[12392]:  cu-4  Job: 58185004 Solution: 0xa51dfaf9eb091a24
Feb 27 14:02:30 gpurig3 bash[12392]:  miner **Accepted  51 ms. us1.ethermine.org:5555
```

This is the output from nvidia-smi for the card
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.32.03    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  GeForce RTX 306...  On   | 00000000:01:00.0 Off |                  N/A |
| 70%   66C    P2   129W / 130W |   4416MiB /  7982MiB |    100%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
```

I am currently running around 66C, which I think is an okay temp.  What temp does your card normally run at?

I'm going to try upgrading to the 460.39 driver again and see if I can build with your GCC version.  I've only gone as high as 9.3.  If that doesn't work, then I don't know what the issue is.

**Comment by sonicrules1234** [comment]
@ahnick Looks like mine is running at 60C when mining.  I had tried using nvidia-settings from cli before and the settings didn't actually apply until I used the ones from reddit. I see you got an accepted share at the end of your log there, so at least we made some progress.  I also had to set the percentage for each fan separately in my commands, so that's probably why you're running at 66C instead of 60.

**Comment by ahnick** [comment]
@sonicrules1234 Thought of a couple other questions:

When nsfminer starts for you do you see that the card is using Compute 8.6?  Something like the top line below?
```
Feb 27 14:21:16 gpurig3 bash[2000]:  cu-0  Using Pci 0000:01:00.0: GeForce RTX 3060 Ti (Compute 8.6) Memory : 7.80 GB
Feb 27 14:21:16 gpurig3 bash[2000]:  cu-1  Using Pci 0000:02:00.0: GeForce GTX 1070 (Compute 6.1) Memory : 7.93 GB
Feb 27 14:21:16 gpurig3 bash[2000]:  cu-2  Using Pci 0000:04:00.0: GeForce GTX 1070 (Compute 6.1) Memory : 7.93 GB
Feb 27 14:21:16 gpurig3 bash[2000]:  cu-3  Using Pci 0000:05:00.0: GeForce GTX 1070 (Compute 6.1) Memory : 7.93 GB
Feb 27 14:21:16 gpurig3 bash[2000]:  cu-4  Using Pci 0000:07:00.0: GeForce GTX 1070 (Compute 6.1) Memory : 7.93 GB
Feb 27 14:21:16 gpurig3 bash[2000]:  cu-5  Using Pci 0000:08:00.0: GeForce GTX 1070 (Compute 6.1) Memory : 7.93 GB
Feb 27 14:21:16 gpurig3 bash[2000]:  miner Epoch : 398 Difficulty : 4.00 Gh
```

Also, how long does DAG generation take for you?  Mine takes around 14 seconds for the 3060 Ti right now.  I get the distinct feeling that the DAG is not getting generated correctly, but as to why that would be I don't know.

**Comment by sonicrules1234** [comment]
> @sonicrules1234 Thought of a couple other questions:
> 
> When nsfminer starts for you do you see that the card is using Compute 8.6? Something like the top line below?
> 
> ```
> Feb 27 14:21:16 gpurig3 bash[2000]:  cu-0  Using Pci 0000:01:00.0: GeForce RTX 3060 Ti (Compute 8.6) Memory : 7.80 GB
> Feb 27 14:21:16 gpurig3 bash[2000]:  cu-1  Using Pci 0000:02:00.0: GeForce GTX 1070 (Compute 6.1) Memory : 7.93 GB
> Feb 27 14:21:16 gpurig3 bash[2000]:  cu-2  Using Pci 0000:04:00.0: GeForce GTX 1070 (Compute 6.1) Memory : 7.93 GB
> Feb 27 14:21:16 gpurig3 bash[2000]:  cu-3  Using Pci 0000:05:00.0: GeForce GTX 1070 (Compute 6.1) Memory : 7.93 GB
> Feb 27 14:21:16 gpurig3 bash[2000]:  cu-4  Using Pci 0000:07:00.0: GeForce GTX 1070 (Compute 6.1) Memory : 7.93 GB
> Feb 27 14:21:16 gpurig3 bash[2000]:  cu-5  Using Pci 0000:08:00.0: GeForce GTX 1070 (Compute 6.1) Memory : 7.93 GB
> Feb 27 14:21:16 gpurig3 bash[2000]:  miner Epoch : 398 Difficulty : 4.00 Gh
> ```
> 
> Also, how long does DAG generation take for you? Mine takes around 14 seconds for the 3060 Ti right now. I get the distinct feeling that the DAG is not getting generated correctly, but as to why that would be I don't know.

I also have the same compute 8.6. As for the dag, the latest commit seems to show more information about it and it seemed to generate faster than before for me at 12.8 seconds.  The hashrate seems to vary more than before though.

**Comment by ahnick** [comment]
>  I had tried using nvidia-settings from cli before and the settings didn't actually apply until I used the ones from reddit. I also had to set the percentage for each fan separately in my commands, so that's probably why you're running at 66C instead of 60.

My guess as to why setting the nvidia-settings originally from the CLI didn't work for you was b/c a lot of the instructions for setting the clock and memory offsets for the GTX 1070 series and others always used performance level 3. So instead of nvidia-settings -c :0 -a '[gpu:0]/GPUGraphicsClockOffset[**3**]=-500 you have to change it to nvidia-settings -c :0 -a '[gpu:0]/GPUGraphicsClockOffset[**4**]=-500.  I know this because initially I ran the command the same way as I had done it for my GTX 1070s and had also noticed no effect, but once I changed the 3 to a 4 it worked for me.

Now interestingly, I just tried to apply the nvidia-settings as listed on the reddit page and it did not work for me.  I got the following:
```
$ nvidia-settings -a "[gpu:0]/GPUGraphicsClockOffsetAllPerformanceLevels=-500"


ERROR: Error assigning value -500 to attribute 'GPUGraphicsClockOffsetAllPerformanceLevels' (gpurig3:0[gpu:0]) as specified in assignment
       '[gpu:0]/GPUGraphicsClockOffsetAllPerformanceLevels=-500' (Unknown Error).
```

I wonder if this is not available in 460.32?  I'm going to upgrade now to 460.39 again and give it another go.  Thanks for confirming the Compute 8.6 setting.  At least I feel a bit better that I'm building nsfminer correctly.

**Comment by ahnick** [comment]
@sonicrules1234 Another question for you.  Do you use riser cards and if so what kind/version?  I'm wondering if there could be some incompatibility with the newer 30XX cards and the riser cards I'm using.

**Comment by sonicrules1234** [comment]
> @sonicrules1234 Another question for you. Do you use riser cards and if so what kind/version? I'm wondering if there could be some incompatibility with the newer 30XX cards and the riser cards I'm using.

I'm not using any risers, just plugged it right into the pcie x16.

**Comment by sonicrules1234** [comment]
@ahnick Did you ever get it working?

**Comment by ahnick** [comment]
@sonicrules1234 No, I have even gone as far as installing Windows 10 Home and trying to get the card working for mining with ethminer, but that did not work either.  It shows up fine, but as soon as it starts to do work then I receive errors.  Prior to installing Windows, I had removed the riser and plugged the board into the PCI slot directly under Ubuntu and it didn't work, but I continued to receive the same errors.  I'm going to try one final time under Windows 10 Home to plug the board directly in without the riser card, but if that doesn't work I'm just going to RMA the board.

**Comment by lesjokolat** [comment]
I really think its related to just 4 gigs system memory. Any chance you can add another 4 gigs to temporarily try and see if it resolves itself?

**Comment by ahnick** [comment]
I'm not sure why it would be related to system memory.  I had 6 GeForce GTX 1070s in this rig before and just replaced one of the cards with a 3060 Ti.  I never had an issue with system memory prior when it was just the 6 1070s and even once the 3060 Ti was swapped out I did observe that the DAG would get built out under nsfminer and ethminer for the 3060 Ti (it would take around 14 seconds) it's just the shares that would come back from it were invalid.  

Also, when I initially setup Windows 10 Home,  the Virtual Memory was set to around 10 GB and I was getting OOM errors and once I increased that to 32 GB of virtual memory those memory issues were resolved.  I would have expected that if it was a memory issue those errors would have continued to persist.  The errors I get back are during CUDA execution, which seems to imply that for whatever reason the calculation is being performed incorrectly (typically attributed to aggressive overclock values) and I don't see how a lack of system memory would be effecting that.

**Comment by ahnick** [comment]
Unfortunately, I have verified that this graphics card is defective.  I loaded it up as the only card in the system and had it drive the display.  When attempting to play some games or generally just navigate around, the card produced visual tearing of images/video, random artifacts being rendered, occasional flickering with black screens, and UI elements not being rendered or rendered incorrectly.  I do appreciate all the assistance provided to help troubleshoot the card.  I'm closing this issue.
